src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -891,18 +878,30 @@ private bool TryAddInternal(TKey key, int hashcode, TValue value, bool updateIfE
                 TValue value;
                 if (!TryGetValue(key, out value))
                 {
-                    throw new KeyNotFoundException();
+                    ThrowKeyNotFoundException();
                 }
                 return value;
             }
             set
             {
-                if (key == null) throw new ArgumentNullException("key");
+                if (key == null) ThrowKeyNullException();
                 TValue dummy;
                 TryAddInternal(key, _comparer.GetHashCode(key), value, true, true, out dummy);
             }
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Might consider including a comment saying that we don't inline these because we think they are on uncommon paths and we don't want to increase the size of methods that may call them (which would prevent the caller's from being inline candidates).